### PR TITLE
More array splitting, handle select and phi

### PR DESCRIPTION
**Pull request by kripken**
For #178

### CODE REVIEW

File: lib/Target/JSBackend/NaCl/ExpandStructRegs.cpp [hunk 1]
```diff
@@ -319,6 +319,81 @@ static bool SplitUpArrayLoad(LoadInst *Load, const DataLayout *DL) {
...
+      BasicBlock *IncomingBB = Phi->getIncomingBlock(PhiIndex);
+      Value *EV = CopyDebug(
+          ExtractValueInst::Create(Phi->getIncomingValue(PhiIndex), EVIndexes,
+                                   Phi->getName() + ".extract",
+                                   IncomingBB->getTerminator()),
+          Phi);
+      NewPhi->addIncoming(EV, IncomingBB);
+    }
+
+    // Reconstruct the original struct value.
```
Comment (eddyb) [inline lib/Target/JSBackend/NaCl/ExpandStructRegs.cpp]: s/struct/array

File: lib/Target/JSBackend/NaCl/ExpandStructRegs.cpp [hunk 2]
```diff
@@ -319,6 +319,81 @@ static bool SplitUpArrayLoad(LoadInst *Load, const DataLayout *DL) {
   return NeedsAnotherPass;
 }
 
+static bool SplitUpArraySelect(SelectInst *Select) {
+  ArrayType *ATy = cast<ArrayType>(Select->getType());
+  Value *NewStruct = UndefValue::get(ATy);
+
+  bool NeedsAnotherPass = false;
+  // Create a separate select instruction for each struct field.
```
Comment (eddyb) [inline lib/Target/JSBackend/NaCl/ExpandStructRegs.cpp]: s/struct field/array element/

File: lib/Target/JSBackend/NaCl/ExpandStructRegs.cpp [hunk 3]
```diff
@@ -319,6 +319,81 @@ static bool SplitUpArrayLoad(LoadInst *Load, const DataLayout *DL) {
...
+                                                EVIndexes, "", Select);
+    Value *FalseValue = ExtractValueInst::Create(Select->getFalseValue(),
+                                                 EVIndexes, "", Select);
+    SelectInst *NewSelect = SelectInst::Create(Select->getCondition(),
+                                               TrueValue, FalseValue,
+                                               Select->getName(),
+                                               Select);
+    NeedsAnotherPass = NeedsAnotherPass || DoAnotherPass(NewSelect);
+
+    // Reconstruct the struct value.
```
Comment (eddyb) [inline lib/Target/JSBackend/NaCl/ExpandStructRegs.cpp]: s/struct/array

**Comment by eddyb** [comment]
I am a bit surprised that it's not easy to handle load/store/select/phi for both structs and arrays a bit more uniformly, but I suppose C++ and LLVM don't make it easy 😞.

**Comment by kripken** [comment]
Thanks, a bunch of places needed struct=>array renaming, fixed.

And yeah, it's a little sad we can't use a template for both. But we hit things like getElementType() needing an index for structs but not for arrays, so indeed C++ and LLVM make that tricky...

**Pull request merged**